pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.kt

```diff
@@ -0,0 +1,137 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.apex.ast
+
+import net.sourceforge.pmd.annotation.InternalApi
+import net.sourceforge.pmd.lang.apex.ApexParserOptions
+import net.sourceforge.pmd.lang.ast.ParseException
+import net.sourceforge.pmd.lang.ast.SourceCodePositioner
+
+import com.google.summit.ast.CompilationUnit
+import com.google.summit.ast.Identifier
+import com.google.summit.ast.Node
+import com.google.summit.ast.SourceLocation
+import com.google.summit.ast.TypeRef
+import com.google.summit.ast.declaration.ClassDeclaration
+import com.google.summit.ast.declaration.EnumDeclaration
+import com.google.summit.ast.declaration.InterfaceDeclaration
+import com.google.summit.ast.declaration.TriggerDeclaration
+import com.google.summit.ast.declaration.TypeDeclaration
+import com.google.summit.ast.modifier.KeywordModifier
+import com.google.summit.ast.modifier.Modifier
+
+@Deprecated("internal")
+@InternalApi
+@Suppress("DEPRECATION")
+class ApexTreeBuilder(val sourceCode: String, val parserOptions: ApexParserOptions) {
+    private val sourceCodePositioner = SourceCodePositioner(sourceCode)
+
+    /** Builds and returns an [ApexNode] AST corresponding to the given [root] node. */
+    fun buildTree(root: CompilationUnit): ApexRootNode<TypeDeclaration> {
+        // Build tree
+        val result =
+            build(root, parent = null) as? ApexRootNode<TypeDeclaration>
+                ?: throw ParseException("Unable to build tree")
+
+        // Call additional methods
+        callAdditional(result)
+
+        return result
+    }
+
+    /** Calls additional methods for each node in [root] using a post-order traversal. */
+    private fun callAdditional(root: ApexNode<*>) =
+        root.jjtAccept(
+            object : ApexParserVisitorAdapter() {
+                override fun visit(node: ApexNode<*>?, data: Any?): Any? =
+                    super.visit(node, data).also {
+                        if (node is AbstractApexNode<*>) {
+                            node.handleSourceCode(sourceCode)
+                            node.calculateLineNumbers(sourceCodePositioner)
+                        }
+                    }
+            },
+            null
+        )
+
+    /**
+     * Builds an [ApexNode] wrapper for [node].
+     *
+     * Sets the parent of the resulting [ApexNode] to [parent], if it's not `null`.
+     */
+    private fun build(node: Node?, parent: ApexNode<*>?): AbstractApexNode<*>? {
+        val wrapper: AbstractApexNode<*>? =
+            when (node) {
+                null -> null
+                is CompilationUnit -> build(node.typeDeclaration, parent)
+                is TypeDeclaration -> buildTypeDeclaration(node)
```

Comment:
Would it make sense to move the logic required to convert from summit.Node --> ApexNode into the nodes themselves? 
It could be either a builder method on the ApexNode implementations taking a summit.Node, or a method on summit.Node returning the corresponding ApexNode?

I'm not too familiar with this code so I'm not sure what restrictions do you want to have on dependencies: should the summit library "know about" and take a dependency on ApexNode? ApexNode has already got a dependency on summit (for the template parameter <T> but my understanding is that that might also go away in the future.

Reply:
Thanks for the suggestion. @aaronhurst-google I would like to hear your thoughts on this as well. Here are mine:

I think the most appropriate location for conversion logic, if not the tree builder class, would be on each Summit node, with each branch of the `when` expression corresponding to the conversion implementation by one Summit node.

Since the node conversions implemented so far are 1:1, this is harder to see, but placing conversion logic in `ApexNode`s would be less effective in my opinion. Here is a more concrete example using another node type:
```kotlin
private fun buildUnaryExpression(node: UnaryExpression) =
  when (node.op) {
    UnaryExpression.Operator.PLUS,
    UnaryExpression.Operator.NEGATION,
    UnaryExpression.Operator.LOGICAL_COMPLEMENT,
    UnaryExpression.Operator.BITWISE_NOT,
    UnaryExpression.Operator.PRE_INCREMENT,
    UnaryExpression.Operator.PRE_DECREMENT, -> ASTPrefixExpression(node)
    UnaryExpression.Operator.POST_INCREMENT,
    UnaryExpression.Operator.POST_DECREMENT, -> ASTPostfixExpression(node)
  }.apply { buildChildren(node, parent = this) }
```
If the conversion logic was placed in `UnaryExpression`, all this internal logic before returning an `ApexNode` instance would be encapsulated properly. If the conversion logic was in `ASTPrefixExpression` and `ASTPostfixExpression`, all that they could hold is a simple function returning a new node instance (everything after the `->`), while the rest of the internal logic is not captured and has to remain in the tree builder, which is what we are trying to avoid. There are many instances like this of one Summit node mapping to many PMD nodes.

However, since Summit AST is its own library separate from PMD, it definitely shouldn't "know about" `ApexNode` or other PMD classes directly. In this situation, I immediately think of the visitor pattern, which could be used here to assign additional behaviors to each Summit node without modifying the library directly. One issue though is that the [`Visitor`](https://github.com/google/summit-ast/blob/main/src/main/java/com/google/summit/ast/traversal/Visitor.kt) that's already implemented in Summit doesn't distinguish between node types (e.g. like [`ApexParserVisitor`](https://github.com/pmd/pmd/blob/master/pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexParserVisitor.java) does). I'm not sure if it would be worth changing or if somehow the node types can still be distinguished without a large `when`.

Regardless, I'm not sure if the visitor solution that distinguishes between nodes is a much better design than the current tree builder anyway. All the conversion logic would still end up in the tree builder (since it can't go in the nodes), and in both cases the work is distributed between many functions. The only effective difference I can see is that the `instanceof`/`is` checks will be replaced with polymorphic checks.

So, if we could modify Summit nodes to contain this logic, I'd definitely prefer that design over the large tree builder we will have in the end, but this isn't really possible. The alternatives that I mentioned don't seem preferable to what we have now.

Reply:
Right, the subtree:subtree mapping from Summit to PMD makes it a difficult to partition the logic into the output classes. And I agree that making PMD a dependency of Summit is backward.

Too bad that Kotlin extension functions have static and not virtual dispatch, otherwise that could have been a solution. (Although +1 for Swift, which allows you to extend the protocols (interfaces) implemented by a class.)